pandas/core/strings.py

```diff
@@ -2280,6 +2280,14 @@ def cat(self, others=None, sep=None, na_rep=None, join=None):
                                  'must all be of the same length as the '
                                  'calling Series/Index.')
 
+        if any(not (x.dtype == 'O' or (is_categorical_dtype(x)
+                                       and x.cat.categories.dtype == 'O')
+                    or (x.dtype == 'float' and x.isna().all()))
```

Comment:
This logic looks somewhat complicated.  Add a comment explaining why we need this type of check.

Reply:
even though it's now documented, i think i would still prefer to see this split somehow.

Reply:
Will try.